packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx

```diff
@@ -0,0 +1,174 @@
+import BigNumber from 'bignumber.js'
+import { Typography } from '@mui/material'
+import { useState, useMemo } from 'react'
+import { useAsync } from 'react-use'
+import { unreachable } from '@dimensiondev/kit'
+import {
+    EthereumTokenType,
+    useNativeTokenDetailed,
+    useFungibleTokenBalance,
+    useWeb3,
+    useAccount,
+    formatBalance,
+} from '@masknet/web3-shared-evm'
+import { TokenAmountPanel } from '@masknet/shared'
+import { useI18N } from '../../../utils'
+import { useStyles } from './SavingsFormStyles'
+import { IconURLS } from './IconURL'
+import { TabType, ProtocolType } from '../types'
+import { SavingsProtocols } from '../protocols'
+import { isLessThan, rightShift } from '@masknet/web3-shared-base'
+import { EthereumWalletConnectedBoundary } from '../../../web3/UI/EthereumWalletConnectedBoundary'
+import { EthereumChainBoundary } from '../../../web3/UI/EthereumChainBoundary'
+import { ActionButtonPromise } from '../../../extension/options-page/DashboardComponents/ActionButton'
+
+export interface SavingsFormProps {
+    chainId: number
+    selectedProtocol: ProtocolType
+    tab: TabType
+    onClose?: () => void
+    onSwapDialogOpen?: () => void
+}
+
+export function SavingsForm({ chainId, selectedProtocol, tab, onClose, onSwapDialogOpen }: SavingsFormProps) {
+    const { t } = useI18N()
+    const { classes } = useStyles()
+    const protocol = SavingsProtocols[selectedProtocol]
+    const targetChainId = chainId
+
+    const { value: nativeTokenDetailed } = useNativeTokenDetailed()
+    const web3 = useWeb3(false, targetChainId)
+    const account = useAccount()
+
+    const [inputAmount, setInputAmount] = useState('')
+    const [estimatedGas, setEstimatedGas] = useState<BigNumber.Value>(new BigNumber('0'))
+
+    const { value: nativeTokenBalance } = useFungibleTokenBalance(EthereumTokenType.Native, '', targetChainId)
+
+    //#region form variables
+    const tokenAmount = new BigNumber(rightShift(inputAmount || '0', 18))
+    const inputAsBN = new BigNumber(rightShift(inputAmount, 18))
+    const balanceAsBN = new BigNumber(TabType.Deposit ? nativeTokenBalance || '0' : protocol.balance)
+
+    useAsync(async () => {
+        const gasEstimate =
+            tab === TabType.Deposit
+                ? await protocol.depositEstimate(account, targetChainId, web3, new BigNumber('1'))
+                : await protocol.withdrawEstimate(account, targetChainId, web3, new BigNumber('1'))
+        setEstimatedGas(gasEstimate)
```

Comment:
Please make it returns the gasEstimate directly, `useState` is redundancy here.

Reply:
@guanbinrui,using the contract method. `estimateGas` uses:

```typescript
(method) PayableTransactionObject<string>.estimateGas(tx?: PayableTx | undefined): Promise<number>
```

Unfortunately there's no way to use it without async.